src/game/logic/object/locomotor.cpp

```diff
@@ -1192,28 +1424,373 @@ void Locomotor::Move_Towards_Position_Wings(
+        case Z_FIXED_ABSOLUTE_HEIGHT: {
+            ret = true;
+            Coord3D pos = *obj->Get_Position();
+            bool relative = m_template->m_behaviorZ == Z_FIXED_SURFACE_RELATIVE_HEIGHT;
+            float surface_at_pt;
+
+            if (relative) {
+                surface_at_pt = Get_Surface_Ht_At_Pt(pos.x, pos.y);
+            } else {
+                surface_at_pt = 0.0f;
+            }
+
+            pos.z = surface_at_pt + m_preferredHeight;
+            obj->Set_Position(&pos);
+            break;
+        }
+        case Z_FIXED_RELATIVE_TO_GROUND_AND_BUILDINGS: {
+            ret = true;
+            Coord3D pos = *obj->Get_Position();
+            pos.z = g_thePartitionManager->Get_Ground_Or_Structure_Height(pos.x, pos.y) + m_preferredHeight;
+            obj->Set_Position(&pos);
+            break;
+        }
+        case Z_RELATIVE_TO_HIGHEST_LAYER: {
+            ret = true;
+            if (m_preferredHeight != 0.0f || Get_Flag(PRECISE_Z_POS)) {
+                Coord3D pos = *obj->Get_Position();
+                PathfindLayerEnum layer = obj->Get_Layer();
+
+                if (layer == LAYER_GROUND) {
+                    layer = g_theTerrainLogic->Get_Highest_Layer_For_Destination(&pos, false);
+                }
+
+                Coord3D pos2;
+                float surface_at_pt = g_theTerrainLogic->Get_Layer_Height(pos.x, pos.y, layer, &pos2, false);
+                float preferred_height = surface_at_pt + m_preferredHeight;
+
+                if (Get_Flag(PRECISE_Z_POS)) {
+                    preferred_height = goal_pos.z;
+                }
+
+                float f = preferred_height - pos.z;
+                f = Get_Preferred_Height_Damping() * f;
+                preferred_height = pos.z + f;
+                float lift = Calc_Lift_To_Use_At_Pt(obj, physics, pos.z, surface_at_pt, preferred_height);
+
+                if (lift != 0.0f) {
+                    Coord3D force;
+                    force.x = 0.0f;
+                    force.y = 0.0f;
+                    force.z = physics->Get_Mass() * lift;
+                    physics->Apply_Motive_Force(&force);
+                }
+            }
+
+            break;
+        }
+    }
+
+    return ret;
+}
+
+float Locomotor::Calc_Lift_To_Use_At_Pt(
+    Object *obj, PhysicsBehavior *physics, float cur_z, float surface_at_pt, float preferred_height)
+{
+    float max_lift = Get_Max_Lift(obj->Get_Body_Module()->Get_Damage_State());
+    float lift = max_lift + g_theWriteableGlobalData->m_gravity;
+
+    if (lift < 0.0f) {
+        lift = 0.0f;
+    }
+
+    float z = physics->Get_Velocity().z;
+    float f1;
+
+    if (Get_Flag(ULTRA_ACCURATE)) {
+        if (z < 0.0f) {
+            f1 = 2.0f * lift;
+        } else {
+            f1 = -2.0f * lift;
+        }
+    } else if (z < 0.0f) {
+        f1 = lift;
+    } else {
+        f1 = g_theWriteableGlobalData->m_gravity;
+    }
+
+    float f3;
+
+    if (GameMath::Fabs(f1) > 0.001f) {
```

Comment:
use GAMEMATH_EPSILON

Reply:
GAMEMATH_EPSILON is 0.0001f, this is 0.001f. Would rather keep things matching stock to prevent hard-to-debug issues later...